docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+
+#: description="Set to deploy containerized node"
+r_openshift_node_is_containerized: false
+#: description="Set to deploy a node over AH"
+r_openshift_node_is_atomic: false
+```
+
+The role variables can be broken down intro three categories:
+* required role variables but undefined: role variables that must be always defined but without a default value
+* optional role variables but undefined: role variables that are undefined by default but can be defined
+* role variables that are always defined with a default value: role variables that can be overwritten
+
+All three categories can be annotated with the following keywords:
+* name: optional annotation used for role variables that are undefined, used when a table of role variables and pre-checks are generated
+* description: short and accurate description of a role variable
+* required: set to `false` if not specified, if set to `true` the role variable is required (forced by generated pre-checks tasks)
+* choices: a list of acceptable values of a role variables
+* required_when: a list of conjuncts under which a role variable must be defined
+
+A table of role variables and a sequence of pre-checks are generated from the specified annotations.
+
+All variables under `var/main.yml` are considered local variables.
+
+#### We should check all relevant role variables are properly set
+
+A list of tasks checking role variables can be generated based on the role variable annotations.
+Given a list of role variables and their properties (e.g. if a variable is required, a list of accepted values)
+can change over time, it is important to keep the checks in sync with the `default/main.yml`.
+The following checks can be generated:
+
+```yaml
+- name: "Fail if r_openshift_node_deployment_type is not defined"
+  fail:
+    msg: "r_openshift_node_deployment_type must be specified for this role"
+  when:
+  - r_openshift_node_deployment_type is not defined
+
+- name: Fail if invalid r_openshift_node_deployment_type provided
+  fail:
+    msg: "r_openshift_node_deployment_type can only be set to a single value from ['origin', 'openshift-enterprise']"
+  when:
+  - r_openshift_node_deployment_type is defined
+  - r_openshift_node_deployment_type not in ['origin', 'openshift-enterprise']
+```
+
+All the checks can be stored into `tasks/pre_checks.yml` file and included in the main task file.
+
+#### Propagation of values to role dependencies
+
+A role can depend on another role that has its own set of role variables. There are two ways how to propagate values to role dependencies. First approach requires a caller has a knowledge of all role dependency variables:
```

Comment:
A role should not utilize variables not defined inside the role whenever possible.
eg, inside the role, role_x, all variables are role_x_something, role_x_something_else.  Never role_y_something inside of role_x.  If role_x and role_y need the same value for some variable, those variables should be set in inventory or elsewhere individually, possibly pointing to the same meta variable.

If role_y sets a fact that role_x needs to consume, and the only way to have that data is via role_x consuming role_y's variable directly, we should consider if we in fact need two separate roles, as we probably need one role.